mollie/api/resources/base.py

```diff
@@ -72,6 +72,9 @@ def list(self, **params):
 
     def perform_api_call(self, http_method, path, data=None, params=None):
         resp = self.client.perform_http_call(http_method, path, data, params)
+        if 'application/hal+json' in resp.headers.get('Content-Type', ''):
+            # set the content type according to the media type definition
+            resp.encoding = 'utf-8'
```

Comment:
For my own understanding, so this is necessary to properly decode unicode bytes? I feel like the comment is wrong 🤔

Reply:
The problem is that the `requests` library decodes the response based on the content-type that the HTTP response header defines (in the `charset` field). The Mollie API doesn't specify a charset, so `requests` falls back to the default which is ISO-8859-1. Decoding the JSON with ISO-8859-1 will fail if there are unicode characters.

The RFC that originally defined the ISO-8859-1 default 
(see https://tools.ietf.org/html/rfc2616#section-3.7.1), is superseded by other RFCs, and the new default is "The default charset of ISO-8859-1 for text media types has been
   removed; the default is now whatever the media type definition says." (see https://tools.ietf.org/html/rfc7231#appendix-B).

As JSON is defined to always be in unicode, and in a perfect world, `requests` would be able to detect this based on the `Content-Type: application/hal+json` header, but it doesn't. Technically this might not even apply, since the RFC only removes the default in text media types; JSON is an application media type.

In a different perfect world, Mollie would send a `Content-Type: application/hal+json charset=utf-8` header, which is recognized by `requests` and removes confusion about which unicode charset would be used. We communicated about this with @ndijkstra and @willemstuursma, unfortunately without any response from their side.

Since Mollie defines their JSON responses to be in UTF-8, we decided to hardcode the value here based on the media type. Feel free to try and fix the Mollie server responses, and we'd be happy to remove the hardcoded value.

Reply:
I'll throw it in the team, and see what comes out. For now we can merge this 👍

Reply:
Since it is an application media type, it doesn't have to specify an encoding does it? Same as for `application/pdf`. JSON is a way of mapping objects to _bytes_, so the character encoding doesn't really matter (that it uses UTF-8 internally is of no concern to the HTTP protocol). Its all handled by the JSON decode (application) not the HTTP transport layer. 

Personally, I think the auto converting behavior is a bit weird on `application/*` content types. 

Regardless, I think the change by @whyscream is valuable. Its probably something many Python implementations might run into.

Reply:
I think that this is exactly the problem: it doesn't matter for HTTP, but the server should use the HTTP header (i.e. metadata) to describe the contents of the response. It does this by saying it's 'application/hal+json'.

Unfortunately, `requests` doesn't use this, so implementors need to apply their own logic based on the content type (you could say that `requests` tries to do a best effort by using the obsoleted `charset=` method, there is a long and still open ticket referenced in the commit about this) until a proper fix within `requests` is available.

Anyway: sending `charset=utf-8` as part of the Content-Type might aid others, but won't hurt anyone.
Some of the Mollie docs show actually document that they send this (f.i. https://docs.mollie.com/reference/v2/payments-api/create-payment#id5), but the production API currently doesn't send it.

I also mentioned in earlier messages that actually sending utf-8 responses makes things complicated for no reason. Why can I create a payment in Euro, but only when I use an ISO 4217 code, but is Mollie allowed to return me an actual Euro sign, when it would be a lot easier when 
I would get the same ISO code in return? ;)

Reply:
It is also possible that UTF-8 data is returned as part of the payment description or in other parts of the API (e.g. organizations/me or profiles/me), so it is impossible to avoid.